Rspec環境構築

pokotyan opened the pull request:
# WHAT

rspec-railsを用いたテストのための環境構築
# WHY

テストを行うため

pokotyan committed:
create pull request

pokotyan committed:
install gems for test

pokotyan committed:
setting change

--format documentation
　rspecの出力をデフォルトから読みやすいドキュメント形式に変更

config.include FactoryGirl::Syntax::Methods
　テストを書く際、FactoryGirlの記述を省略するため

Capybara.javascript_driver = :poltergeist
　Capybaraの標準のwebドライバは、Rack::Testというもので、このドライバはJavaScriptを実行しないため、JavaScriptを無視されます。
　Javasciptを使用しているProtospaceにはJavaScriptの実行可能なwebドライバが必要となります。
　そこで、JavaScriptに対応し、かつheadless(webブラウザを起動することなくテストを行う)なpoltergeistというgemを使用します。

config.use_transactional_fixtures = false
　この設定がtrueになっている場合、すべてのテストはトランザクション中で実行される。
　なので、create(:user)もトランザクション中で実行される。
　Capybaraでvisit root_pathしても、create(:user)のトランザクションはコミットされていないため、レコードが作成されておらず、テスト失敗の原因になる。
　use_transactional_fixturesをfalseにするとレコードが作成されるようになる。
　ただし、その場合はテストで作成したレコードを削除するなどの処理を自分で管理する必要があるため、
　database_cleanerをつかい、テストで作成したレコードを削除する

config.before(:suite) do
　DatabaseCleaner.clean_with(:truncation)
テストスイート全体が実行される前にテストデータベースを完全に消去します。
これにより、中断したテストや下手なテストコードによって生じたゴミを一掃（トランケート）し、テストが思わぬ振る舞いをするのを防ぎます。

DatabaseCleaner.strategy = :transaction
　上の設定は、データベース消去のストラテジを”transaction”に設定します。トランザクションは非常に速いので、
　トランザクションを実行するすべてのテスト(実際のところ、RSpecプロセスで実行されるすべてのテスト)において、この設定を行っておくのがよいでしょう。

config.around(:each) do |example|
    DatabaseCleaner.cleaning do
      example.run
トランザクションの開始と終了を実行するタイミング(各 example を実行する前後)でデータ削除

kenchan0503 commented on spec/rails_helper.rb:
```diff
@@ -0,0 +1,73 @@
+      example.run
+    end
+  end
+end
```
ファイル生成時に存在するコメントアウトは不要なので消しておきましょう。

pokotyan replied:
消しました！

kenchan0503 replied:
👍

kenchan0503 commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,99 @@
+  # as the one that triggered the failure.
+  Kernel.srand config.seed
+=end
+end
```
こちらも同様です。

pokotyan replied:
消しました！

kenchan0503 replied:
👍

kenchan0503 commented on spec/rails_helper.rb:
```diff
@@ -0,0 +1,73 @@
+  config.before(:suite) do
+    DatabaseCleaner.strategy = :transaction
+    DatabaseCleaner.clean_with(:truncation)
+  end
```
database_cleanerの設定は新たにファイルを生成してそちらに記述した方がいいですね！
rails_helper内の記述は全てのspecファイルに共通する内容ですが、database_cleanerはspecファイルの全てで使用するわけではないので。

kenchan0503 replied:
http://master.tech-camp.in/curriculums/708
詳しくはこちらのカリキュラムに記載されているのでこちらを参考にしてみてください！！

pokotyan replied:
別ファイルにdatabase_cleanerの設定の記述を移しました！

kenchan0503 replied:
👍

kenchan0503 commented on spec/rails_helper.rb:
```diff
@@ -0,0 +1,73 @@
+      example.run
+    end
+  end
+end
```
こちらも同様です。

pokotyan replied:
移しました！

kenchan0503 replied:
👍

pokotyan committed:
delete comment-out and Move database-cleaner.rb

rails_helper内の記述は全てのspecファイルに共通する内容だが、database_cleanerはspecファイルの全てで使用するわけではないので、database_cleanerの記述を別ファイルに移動。

kenchan0503 commented:
LGTMです！！
